include/alpaka/block/shared/st/BlockSharedMemStMember.hpp

```diff
@@ -0,0 +1,146 @@
+#ifndef NDEBUG
+                        BlockSharedMemStMemberImpl(uint8_t* mem, unsigned int capacity) : m_mem(mem), m_capacity(capacity) {}
+#else
+                        BlockSharedMemStMemberImpl(uint8_t* mem, unsigned int) : m_mem(mem) {}
+#endif
+                        //-----------------------------------------------------------------------------
+                        BlockSharedMemStMemberImpl(BlockSharedMemStMemberImpl const &) = delete;
+                        //-----------------------------------------------------------------------------
+                        BlockSharedMemStMemberImpl(BlockSharedMemStMemberImpl &&) = delete;
+                        //-----------------------------------------------------------------------------
+                        auto operator=(BlockSharedMemStMemberImpl const &) -> BlockSharedMemStMemberImpl & = delete;
+                        //-----------------------------------------------------------------------------
+                        auto operator=(BlockSharedMemStMemberImpl &&) -> BlockSharedMemStMemberImpl & = delete;
+                        //-----------------------------------------------------------------------------
+                        /*virtual*/ ~BlockSharedMemStMemberImpl() = default;
+
+                        template <typename T>
+                        void alloc() const
+                        {
+                            uint8_t* buf = &m_mem[m_allocdBytes];
+                            new (buf) T();
+                            m_allocdBytes += alignPitch<T>();
+#if (defined ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST) && (! defined NDEBUG)
+                            ALPAKA_ASSERT(m_allocdBytes < m_capacity);
+#endif
+                        }
+
+#if BOOST_COMP_GNUC
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align" // "cast from 'unsigned char*' to 'unsigned int*' increases required alignment of target type"
+#endif
+                        template <typename T>
+                        T& getLatestVar() const
+                        {
+                           return *reinterpret_cast<T*>(&m_mem[m_allocdBytes-alignPitch<T>()]);
+                        }
+#if BOOST_COMP_GNUC
+    #pragma GCC diagnostic pop
+#endif
+
+                        void free() const
+                        {
+                            m_allocdBytes = 0u;
+                        }
+
+                    };
+                }
+                //#############################################################################
+                //! Static block shared memory provider using a pointer to
+                //! externally allocated fixed-size memory, likely provided by
+                //! BlockSharedMemDynMember.
+                template<unsigned int TDataAlignBytes = 4>
+                class BlockSharedMemStMember :
+                    public detail::BlockSharedMemStMemberImpl<TDataAlignBytes>,
```

Comment:
What is the reason for the inheritance here? Couldn't `BlockSharedMemStMemberImpl` be merged into `BlockSharedMemStMember`?

Reply:
Forgot to mention this in the description, sorry.

I need to derive from Impl for the OpenMP 5 and OpenACC backends. Deriving from `BlockSharedMemStMember`, which already implements the shared mem concept confuses the traits, because then there are two base classes of the accelerator which derive from an instance of the concept.

Reply:
I do not fully get this. Who else is deriving from the same concept and why?

Reply:
Things in the [OpenMP 5](https://github.com/jkelling/alpaka/blob/blockSharedMemberOpenacc/include/alpaka/block/shared/st/BlockSharedMemStOmp5.hpp#L36) and [OpenACC](https://github.com/jkelling/alpaka/blob/blockSharedMemberOpenacc/include/alpaka/ctx/block/CtxBlockOacc.hpp#L53) backends. They need the same small-object allocators (I actually back-ported this code from there), but they need to synchronize threads to call `alloc` and thus need to implement the traits differently.